File: 골드 4 (2문제)/9663_N-Queen.c
// 설명 링크 : https://moon-gd.tistory.com/99
#define _CRT_SECURE_NO_WARNINGS
#define MAX 15
#include <stdio.h>
#include <math.h>

int size = 0, goal = 0;
int columns[MAX];

void queen(int depth);
int validity(int depth);

int main() {
	scanf("%d", &size);

	queen(0);

	printf("%d", goal);

	return 0;
}

void queen(int depth) {
	if (validity(depth)) {
		if (depth == size) {
			goal++;
			return;
		}
		else {
			for (int i = 1; i <= size; i++) {
				columns[depth + 1] = i;
				queen(depth + 1);
			}
		}
	}
}

int validity(int depth) {
	for (int i = 1; i < depth; i++) {
		if (columns[i] == columns[depth] || depth - i == abs(columns[depth] - columns[i])) {
			return 0;
		}
	}

	return 1;
}

File: 골드 4 (1문제)/9663_N-Queen.py
# 설명 링크 : https://moon-gd.tistory.com/99
def validity(depth):
    for i in range(1, depth):
        if depth - i == abs(columns[depth] - columns[i]):
            return False

    return True


def queen(depth):
    if depth == size:
        global goal
        goal = goal + 1
    else:
        for j in range(1, size + 1):
            if visited[j]:
                continue

            columns[depth + 1] = j
            if validity(depth + 1):
                visited[j] = True
                queen(depth + 1)
                visited[j] = False


size = int(input())
goal = 0

columns = [0 for _ in range(size + 1)]
visited = [False for _ in range(size + 1)]

queen(0)
print(goal)
